hypothesis-python/src/hypothesis/internal/conjecture/datatree.py

```diff
@@ -67,61 +97,275 @@ class Conclusion:
+        #
+        #     x**y > n
+        # <=> log(x**y)  > log(n)
+        # <=> y * log(x) > log(n)
+
+        # avoid math.log(1) == 0 and incorrectly failing the below estimate,
+        # even when we definitely are too large.
+        if len(intervals) == 1:
+            definitely_too_large = max_size > MAX_CHILDREN_EFFECTIVELY_INFINITE
+        else:
+            definitely_too_large = max_size * math.log(len(intervals)) > math.log(
+                MAX_CHILDREN_EFFECTIVELY_INFINITE
+            )
+
+        if definitely_too_large:
+            return MAX_CHILDREN_EFFECTIVELY_INFINITE
+
+        # number of strings of length k, for each k in [min_size, max_size].
+        return sum(len(intervals) ** k for k in range(min_size, max_size + 1))
+
+    elif ir_type == "float":
+        return count_between_floats(kwargs["min_value"], kwargs["max_value"])
+    else:  # pragma: no cover
+        raise ValueError(f"unhandled ir_type {ir_type}")
+
+
+# In theory, this is a strict superset of the functionality of compute_max_children;
+#
+#   assert len(all_children(ir_type, kwargs)) == compute_max_children(ir_type, kwargs)
+#
+# In practice, we maintain two distinct implementations for efficiency and space
+# reasons. If you just need the number of children, it is cheaper to use
+# compute_max_children than reify the list of children (only to immediately
+# throw it away).
+def all_children(ir_type, kwargs):
+    if ir_type == "integer":
+        min_value = kwargs["min_value"]
+        max_value = kwargs["max_value"]
+        weights = kwargs["weights"]
+        # it's a bit annoying (but completely feasible) to implement the cases
+        # other than "both sides bounded" here. We haven't needed to yet because
+        # in practice we don't struggled with unbounded integer generation.
+        assert min_value is not None
+        assert max_value is not None
+
+        if weights is None:
+            yield from range(min_value, max_value + 1)
+        else:
+            # skip any values with a corresponding weight of 0 (can never be drawn).
+            for weight, n in zip(weights, range(min_value, max_value + 1)):
+                if weight == 0:
+                    continue
+                yield n
+
+    if ir_type == "boolean":
+        p = kwargs["p"]
+        if p <= 2 ** (-64):
+            yield False
+        elif p >= (1 - 2 ** (-64)):
+            yield True
+        else:
+            yield from [False, True]
+    if ir_type == "bytes":
+        size = kwargs["size"]
+        yield from (int_to_bytes(i, size) for i in range(2 ** (8 * size)))
+    if ir_type == "string":
+        min_size = kwargs["min_size"]
+        max_size = kwargs["max_size"]
+        intervals = kwargs["intervals"]
+
+        size = min_size
+        while size <= max_size:
+            for ords in itertools.product(intervals, repeat=size):
+                yield "".join(chr(n) for n in ords)
+            size += 1
```

Comment:
I wrote this weirdly because there is a `max_size = math.inf` case here with `@given(st.text("a"))`, which delegates to `ListStrategy`, which converts `max_size=None` to `self.max_size = math.inf` which `TextStrategy` then passes to our provider.

There's a few of places we could change this behavior, if desired - setting `self.max_size` in `TextStrategy` or treating `max_size=math.inf` the same as `max_size=None` in `PrimitiveProvider.draw_float` are the two that come to mind. Alternatively, leave this inf-handling-code in place (but this hides a maybe-bug with `max_size=math.inf` overriding our "effectively infinite" 10**10 in draw_string).

Reply:
Ah, that's fine to handle the `inf` case, just write a comment in the code for future readers?